File: clojure/src/ch1/ex124.clj
;; (load-file "filename.clj")
;; or on command line> clj filename.clj


(def *FAST-TIMES* 2)


(defn divides? [a b]
  (= (rem b a) 0))


(defn square [n]
  (* n n))


(defn odd? [n]
  (not (divides? 2 n)))


(defn expmod [base exp m]
  (cond ((= exp 0) 1
         (not (odd? exp))
           (rem (square (expmod base (/ exp 2) m)) m)
         :else 
           (rem (* base (expmod base (dec exp) m)) m))))
  

(defn fermat-test [n]
  (let [try-it (fn [a] (= (expmod a n n) a))]
    (try-it (inc (rand-int (dec n))))))


(defn fast-prime? [n times]
  (cond ((= times 0) true
         (fermat-test n) (fast-prime? n (dec times))
         :else false)))


(defn prime? [n]
  (fast-prime? n *FAST-TIMES*))


(defn prime-test [n]
  (if (prime? n)
    (println n " *** ")
    (println n)))


(defn search-helper [n end]
  (if (<= n end)
    (do (prime-test n)
      (search-helper (+ n 2) end))
    (do (println "finished\n"))))


(defn search-for-primes [start end]
  (search-helper (if (odd? start) start (inc start)) end))


(require
 'clojure.test)

(println (search-for-primes 1000 1020))
(println (search-for-primes 10000 10040))
(println (search-for-primes 100000 100080))

(println "done")

File: haskell/ch1/Ex124.hs
-- :cd C:\Users\Yohan\IdeaProjects\sicp\haskell\ch1
-- :load ExXXX
-- Run with func_name arg1...
module Ex124 where

import System.Random
import Control.Monad
import Data.Char

fast_times = 3

sq n = n * n

divides a b = (mod b a) == 0

rand_nums n = replicateM n (randomRIO (0 :: Double, 1)) 

rand_ints :: Int -> Int -> Int -> IO [Int]
rand_ints n lo hi = do
  nums <- replicateM n (randomRIO (lo :: Int, hi)) 
  return nums

random_num :: Integer -> Integer
random_num limit = do
--  gen <- mkStdGen limit
  limit - 1

find_divisor n test | sq test > n = n
find_divisor n test | divides test n = test
find_divisor n test = find_divisor n (test + 1)

expmod :: Integer -> Integer -> Integer -> Integer
expmod _ 0 _ = 1
expmod base exp m | even exp = rem (sq (expmod base (div exp 2) m)) m
expmod base exp m = rem (base * (expmod base (exp - 1) m)) m

--Could not figure out random number version
fermat_test :: Integer -> Bool
fermat_test n = 
  let try_it a = (expmod a n n) == a
  in try_it (1 + random_num n)

fast_prime n 0 = True
fast_prime n times =
  if fermat_test n
  then fast_prime n (n - 1)
  else False

is_prime n = fast_prime n fast_times

prime_test n | is_prime n = show n ++ " ***"
prime_test n = show n

range start end = [start, (start + 2) .. end]

make_range start end = if odd start
                       then range start end
                       else range (start + 1) end

report vals = mapM_ putStrLn vals

search_for_primes start end = report (map prime_test (make_range start end))
